File: internal/auth/middlewares/authentication.go
//go:build js && wasm
package auth_middlewares

import (
	"net/http"
	jwt "github.com/golang-jwt/jwt/v5"
	. "github.com/kodesettings/dbmodule/v2/internal/core"
	database_repository "github.com/kodesettings/dbmodule/v2/internal/database/repository"
)

type auth struct {
	authentication _authentication `json:"auth"`
}

type _authentication struct {
	device_identifier string `json:"deviceIdentifier"`
}

func (a *auth) __decodeJWTToken(token_string string) string {
	_, err := jwt.Parse(token_string, func(token *jwt.Token) (interface{}, error) {
		return token, nil
	});

	return err.Error()
}

func (a *auth) AuthenticationValidator(w http.ResponseWriter, req *http.Request) {
	deviceIdentifier := req.Header.Get("deviceIdentifier")
	authHeader := req.Header.Get("Authorization")
	if authHeader == "" {
		c := ApiError{/*empty struct, it is assigned in function call*/}
		c.BadRequestError("no authorization token provided")
		return
	}

	var token string = authHeader[:7]; // remove Bearer prefix from token
	var decodedToken string = a.__decodeJWTToken(token);

	if decodedToken == "" { c := ApiError{}; c.BadRequestError("invalid tokn"); return; }

	handler := database_repository.RefreshTokenRepo{}
	_, isUserLoggedIn := handler.FindByDeviceIdentifier(deviceIdentifier);

	if !isUserLoggedIn { c:= ApiError{}; c.InternalError("login again"); return; }
/*
    if (decodedToken.usage === "auth-access") {
      const finalDecodedToken = jwt.verify(
        token,
        tokenInfo.jwtSecret
      ) as JwtPayload;

      if (!finalDecodedToken) {
        throw new BadRequestError("send refresh token");
      }

      return new SuccessResponse("successful authentication").send();
    } else {
      throw new BadRequestError("token is not valid");
    }
*/
}

File: sources/auth/middlewares/authentication.cpp
#include "middlewares.hpp"

void authentication(const std::string req, std::string *resp) {
  std::string authHeader, deviceIdentifier; // TODO: get these from header

  if (authHeader.empty()) {
    *resp = BadRequestError("No authorization token provided").prepare();
    return;
  }

  auto token = authHeader.substr(0, 7); // remove Bearer prefix from token
  auto decodedToken = jwt::decode(token);

  if (decodedToken.get_payload_json().empty()) {
    *resp = BadRequestError("Invalid Token").prepare();
    return;
  }

  // TODO: overwrite request
  std::string userId = decodedToken.get_payload_json()["user"].get("id").to_str();

  bool isUserLoggedIn;
  findRefreshTokenByDeviceIdentifier(deviceIdentifier, &isUserLoggedIn);

  if (!isUserLoggedIn) {
    *resp = ForbiddenError("Login Again").prepare();
    return;
  }

  if (decodedToken.get_payload_json()["user"].get("usage").to_str() == "auth-access") {
    try {
      auto verify = jwt::verify();
      verify.verify(decodedToken);
    } catch (...) {
      *resp = BadRequestError("send refresh token").prepare();
    }
  } else {
    *resp = BadRequestError("token is not valid").prepare();
  }
}
